### PR TITLE
[RISCV][GlobalISel] Create generic instructions to handle legalization and selection of G_GLOBAL_VALUE for medlow code model

### PR DESCRIPTION
**Pull request by nitinjohnraj**
We create generic instructions G_RISCV_HI and G_RISCV_ADD_LO. We lower G_GLOBAL_VALUE to these instructions during legalization, and we use pre-existing tablegen rules to select these instructions.

### CODE REVIEW

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp [hunk 1]
```diff
@@ -164,6 +164,25 @@ bool RISCVInstructionSelector::select(MachineInstr &MI) {
   preISelLower(MI, MIB, MRI);
   const unsigned Opc = MI.getOpcode();
 
+  switch(Opc) {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:167]: Why can't this be part of the normal switch?

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp [hunk 2]
```diff
@@ -164,6 +164,25 @@ bool RISCVInstructionSelector::select(MachineInstr &MI) {
...
+  case RISCV::G_RISCV_ADD_LO: {
+    Register DstReg = MI.getOperand(0).getReg();
+    const LLT XLenLLT = LLT::scalar(STI.getXLen());
+    RBI.constrainGenericRegister(DstReg, RISCV::GPRRegClass, MRI);
+    MI.setDesc(TII.get(RISCV::ADDI));
+    MRI.setType(DstReg, XLenLLT);
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:173]: Why do we need to call setType on a virtual register that's got a register class?

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp [hunk 3]
```diff
@@ -248,6 +267,8 @@ bool RISCVInstructionSelector::select(MachineInstr &MI) {
     break;
   }
 
+  errs() << "HERE\n";
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:270]: ???

File: llvm/test/CodeGen/RISCV/GlobalISel/regbankselect/global.mir [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    ret ptr @var
+  }
+
+...
+---
+name:            test_global
+legalized:       true
+tracksRegLiveness: true
+body:             |
+  bb.0 (%ir-block.0):
```
Comment (arsenm) [inline llvm/test/CodeGen/RISCV/GlobalISel/regbankselect/global.mir:22]: drop block name

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 5]
```diff
@@ -176,10 +181,39 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST) {
...
+                        .addGlobalAddress(GV, Offset, RISCVII::MO_HI);
+
+      MIB.buildInstr(RISCV::G_RISCV_ADD_LO, {DstReg}, {AddrHi})
+                      .addGlobalAddress(GV, Offset, RISCVII::MO_LO);
+
+      MI.eraseFromParent();
+      return true;
+    }
+    default:
+      llvm_unreachable("Code model not handled!");
```
Comment (arsenm) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:214]: should prefer to fail legalization by returning false instead of asserting

**Review by arsenm — CHANGES_REQUESTED** [review]

**Pull request closed without merging**